[vulkan] max_pool2d

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42181 [vulkan] Permtute ops: transpose,view,slice,unsqueeze,select
* #42180 [vulkan][op] avg_pool2d
* #42179 [vulkan][op] add.Scalar, mul.Scalar
* #41434 [vulkan] cat op (concatenate)
* #41380 [vulkan] inplace add_, relu_
* **#41379 [vulkan] max_pool2d**

Differential Revision: [D22754944](https://our.internmc.facebook.com/intern/diff/D22754944)

IvanKobzarev committed:
[vulkan] max_pool2d

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] max_pool2d"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] max_pool2d"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] max_pool2d"


Differential Revision: [D22754944](https://our.internmc.facebook.com/intern/diff/D22754944)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] max_pool2d"


Differential Revision: [D22754944](https://our.internmc.facebook.com/intern/diff/D22754944)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] max_pool2d"


Differential Revision: [D22754944](https://our.internmc.facebook.com/intern/diff/D22754944)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] max_pool2d"


Differential Revision: [D22754944](https://our.internmc.facebook.com/intern/diff/D22754944)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] max_pool2d"


Differential Revision: [D22754944](https://our.internmc.facebook.com/intern/diff/D22754944)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] max_pool2d"


Differential Revision: [D22754944](https://our.internmc.facebook.com/intern/diff/D22754944)

[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanAten.cpp:
```diff
@@ -163,6 +164,95 @@ at::Tensor vulkan_adaptive_avg_pool2d(
+  TORCH_CHECK(
+      kernel_size.size() == 1 || kernel_size.size() == 2,
+      "Vulkan max_pool2d: kernel_size must either be a single int, or a tuple of two ints")
+  const int kH = safe_downcast<int, int64_t>(kernel_size[0]);
```
This usage is technically correct but [template argument deduction](https://en.cppreference.com/w/cpp/language/template_argument_deduction) can deduce the second template parameter (i.e. source type) passed as a function argument, so this will work as well:

```
const int kH = safe_downcast<int>(kernel_size[0]); 
```

AshkanAliabadi commented on aten/src/ATen/native/vulkan/VulkanAten.cpp:
```diff
@@ -163,6 +164,95 @@ at::Tensor vulkan_adaptive_avg_pool2d(
+      self.dim() == 4, "Vulkan max_pool2d is implemented for 4-dim input");
+
+  const auto& x = vtensor_from_vulkan(self);
+  auto inputSize = self.sizes();
```
Is this const?  If so auto does not imply const.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/glsl/max_pool2d.glsl:
```diff
@@ -0,0 +1,41 @@
+uConstBlock;
+
+#define UP_DIV(x, y) (((x) + (y)-1) / (y))
+#define FLT_MAX 3.402823466e+38
```
If these are used in other shaders please consider moving to a common location.

AshkanAliabadi reviewed (approved):
Suggestion: more rigorous const-correctness including shaders.

IvanKobzarev committed:
Update on "[vulkan] max_pool2d"


Differential Revision: [D22754944](https://our.internmc.facebook.com/intern/diff/D22754944)

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] max_pool2d"


Differential Revision: [D22754944](https://our.internmc.facebook.com/intern/diff/D22754944)

[ghstack-poisoned]
